src/librbd/operation/EnableFeaturesRequest.cc

```diff
@@ -0,0 +1,444 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/operation/EnableFeaturesRequest.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AioImageRequestWQ.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Journal.h"
+#include "librbd/Utils.h"
+#include "librbd/image/CreateObjectMapRequest.h"
+#include "librbd/image/MirrorEnableRequest.h"
+#include "librbd/image/SetFlagsRequest.h"
+#include "librbd/journal/CreateRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::EnableFeaturesRequest: "
+
+namespace librbd {
+namespace operation {
+
+using util::create_context_callback;
+using util::create_rados_ack_callback;
+
+template <typename I>
+EnableFeaturesRequest<I>::EnableFeaturesRequest(I &image_ctx,
+                                               Context *on_finish,
+                                               uint64_t features)
+  : Request<I>(image_ctx, on_finish), m_features(features) {
+}
+
+template <typename I>
+void EnableFeaturesRequest<I>::send_op() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  assert(image_ctx.owner_lock.is_locked());
+
+  ldout(cct, 20) << this << " " << __func__ << ": features=" << m_features
+		 << dendl;
+
+  if (this->is_canceled()) {
```

Comment:
Minor: no need to check for this -- it's only used for long-running operations (e.g. resize)